src/test/test-dns-domain.c

```diff
@@ -120,6 +120,37 @@ static void test_dns_name_equal(void) {
         test_dns_name_equal_one("..", "..", -EINVAL);
 }
 
+static void test_dns_name_between_one(const char *a, const char *b, const char *c, int ret) {
+        int r;
+
+        r = dns_name_between(a, b, c);
+        assert_se(r == ret);
+
+        r = dns_name_between(c, b, a);
+        if (ret >= 0)
+                assert_se(r == 0);
+        else
+                assert_se(r == ret);
+}
+
+static void test_dns_name_between(void) {
+        /* see https://tools.ietf.org/html/rfc4034#section-6.1 */
+
+        test_dns_name_between_one("example", "a.example", "yljkjljk.a.example", true);
+        test_dns_name_between_one("a.example", "yljkjljk.a.example", "Z.a.example", true);
+        test_dns_name_between_one("yljkjljk.a.example", "Z.a.example", "zABC.a.EXAMPLE", true);
+        test_dns_name_between_one("Z.a.example", "zABC.a.EXAMPLE", "z.example", true);
+        test_dns_name_between_one("zABC.a.EXAMPLE", "z.example", "\001.z.example", true);
+        test_dns_name_between_one("z.example", "\001.z.example", "*.z.example", true);
+        test_dns_name_between_one("\001.z.example", "*.z.example", "\200.z.example", true);
+        test_dns_name_between_one("*.z.example", "\200.z.example", "example", true);
```

Comment:
Right, this needs at least a comment, but maybe it is overly clever and we should simplify it to avoid confusion.

This is the logic: The usecase for the between() function is to check if an NSEC record refutes the existence of a domainname. If a domainname lies between (according to the canonical order) the owner and the next_domain_name of a NSEC record it means the domain name does not exist in the zone.

There is a small quirk in that the last NSEC record of the zone points at the zone apex as its next_domain_name, which is NOT greater than the NSEC owner name, rather it is smaller than all other names in the zone (i.e., it is the 'first' name in the zone). In the between() function, I abstracted this away a bit, and rather than refer to the zone apex, I simply considered the last NSEC record to point to the first name in the zone (which just happens to be the apex), which is the reason for the special handling when the second argument is smaller than the first, in which case we 'wrap around' the zone, and consider the given domain name to be between the other two if it is greater than the last name in the zone, or smaller than the first.

Makes sense?